From 45c3af0ec9e0a162072667771c120d9fca9f921b Mon Sep 17 00:00:00 2001
From: Roman Shaposhnik <rvs@apache.org>
Date: Fri, 22 May 2020 00:26:11 +0000
Subject: [PATCH] Workaround for https://github.com/9fans/plan9port/issues/36

Signed-off-by: Roman Shaposhnik <rvs@apache.org>
---
 bin/9l | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bin/9l b/bin/9l
index b4f9158..c0a84e9 100755
--- a/bin/9l
+++ b/bin/9l
@@ -25,7 +25,7 @@ case "$tag" in
 *Linux*)
 	ld="${CC9:-gcc} $CC9FLAGS"
 	userpath=true
-	extralibs="$extralibs -lutil -lresolv -lpthread"
+	extralibs="$extralibs -lutil -lresolv -lpthread -lucontext"
 	;;
 *Darwin*x86_64*)
 	ld="${CC9:-gcc} -m64 $CC9FLAGS"
-- 
2.24.3

